pkg/virtual/apiexport/authorizer/maximal_permission_policy.go

```diff
@@ -0,0 +1,129 @@
+}
+
+func NewMaximalPermissionAuthorizer(deepSARClient kubernetes.ClusterInterface, kcpClusterClient kcpclient.ClusterInterface) authorizer.Authorizer {
+	return &maximalPermissionAuthorizer{
+		getAPIExport: func(ctx context.Context, clusterName, apiExportName string) (*apisv1alpha1.APIExport, error) {
+			return kcpClusterClient.Cluster(logicalcluster.New(clusterName)).ApisV1alpha1().APIExports().Get(ctx, apiExportName, v1.GetOptions{})
+		},
+		newDelegatedAuthorizer: func(clusterName string) (authorizer.Authorizer, error) {
+			return delegated.NewDelegatedAuthorizer(logicalcluster.New(clusterName), deepSARClient)
+		},
+	}
+}
+
+func (a *maximalPermissionAuthorizer) Authorize(ctx context.Context, attr authorizer.Attributes) (dec authorizer.Decision, reason string, err error) {
+	defer func() {
+		auditReason := reason
+		if err != nil {
+			auditReason = fmt.Sprintf("%v, error: %v", reason, err)
+		}
+		kaudit.AddAuditAnnotations(
+			ctx,
+			VirtualAPIExportMaximumPermissionPolicyAuditDecision, kcpauth.DecisionString(dec),
+			VirtualAPIExportMaximumPermissionPolicyAuditReason, auditReason,
+		)
+	}()
+
+	apiDomainKey := dynamiccontext.APIDomainKeyFrom(ctx)
+	parts := strings.Split(string(apiDomainKey), "/")
+	if len(parts) < 2 {
+		return authorizer.DecisionNoOpinion, VirtualAPIExportNotPermittedReason, fmt.Errorf("invalid API domain key: %q", apiDomainKey)
+	}
+	apiExportCluster := parts[0]
+	apiExportName := parts[1]
+
+	apiExport, err := a.getAPIExport(ctx, apiExportCluster, apiExportName)
+	if err != nil && errors.IsNotFound(err) {
+		return authorizer.DecisionNoOpinion, VirtualAPIExportNotPermittedReason, fmt.Errorf("API export %q not found: %q", apiDomainKey)
+	}
+
+	if apiExport.Spec.MaximalPermissionPolicy == nil {
+		return authorizer.DecisionAllow, "no maximal permission policy defined", nil
+	}
+
+	if apiExport.Spec.MaximalPermissionPolicy.Local == nil {
+		return authorizer.DecisionAllow, "no maximal local permission policy defined", nil
+	}
+
+	prefixedAttr := deepCopyAttributes(attr)
+	userInfo := prefixedAttr.User.(*user.DefaultInfo)
+	userInfo.Name = apisv1alpha1.MaximalPermissionPolicyRBACUserGroupPrefix + userInfo.Name
+	userInfo.Groups = make([]string, 0, len(attr.GetUser().GetGroups()))
+	for _, g := range attr.GetUser().GetGroups() {
+		userInfo.Groups = append(userInfo.Groups, apisv1alpha1.MaximalPermissionPolicyRBACUserGroupPrefix+g)
+	}
+
+	authz, err := a.newDelegatedAuthorizer(apiExportCluster)
+	if err != nil {
+		return authorizer.DecisionNoOpinion, VirtualAPIExportNotPermittedReason, err
+	}
+
+	dec, reason, err = authz.Authorize(ctx, prefixedAttr)
+	if err != nil {
+		return authorizer.DecisionNoOpinion, VirtualAPIExportNotPermittedReason, fmt.Errorf("error authorizing in API export cluster %q: %w", apiExportCluster, err)
+	}
+	reason = fmt.Sprintf("maximal permission policy authorization reason in API export cluster %q: %q", apiExportCluster, reason)
+	return
```

Comment:
nit: am not a fan of non-explicit returns, i.e. via named vars. This also risks that one does forget to set certain vars.

Reply:
honestly, i'm not a fain either. let me see if i can factor this and not have lots of duplication.

Reply:
i'll rework audit here completely